### ISSUE TITLE
Varnish not closing backend connection when client disconnects - waiting for first_byte_timeout instead

### ISSUE THREAD

**Issue by thetuxkeeper** [varnishcache/varnish-cache#3556]
Hi,

we have an issue with connections hanging in the backend and waiting for POST data though the client already closed the connection. This is a quite unexpected behavior (for us) and might overload the backend with not yet closed connections.

Is that a known/expected behavior?

## Expected Behavior
Close the connection to the backend at the time varnish knows the client is gone. Then the backend doesn't wait for data without knowing the client is already gone.

## Current Behavior
read error is logged but first_byte_timeout kills the backend connection much later.
Probably it only occurs with reused/keepalive connections.

## Steps to Reproduce (for bugs)
1. Send a too short POST-Request and close the connection (backend part of log)
```
POST /test HTTP/1.1
Host: test-host
Content-Length: 100000
content-type: application/json

0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
```
2. You'll see something like that in the log /backend part):
```
*   << BeReq    >> 1058444892
-   Begin          bereq 1058444891 pass
-   VCL_use        reload_20210319_140100_14960
-   Timestamp      Start: 1616162470.789302 0.000000 0.000000
-   BereqMethod    POST
-   BereqURL       /test
-   BereqProtocol  HTTP/1.1
-   BereqHeader    content-length: 100000
-   BereqHeader    content-type: application/json
-   BereqHeader    secure: on
-   BereqHeader    x-forwarded-by: ...
-   BereqHeader    x-forwarded-proto: https
-   BereqHeader    x-forwarded-uri: /test
-   BereqHeader    x-forwarded-host: test-host
-   BereqHeader    x-real-ip: ...
-   BereqHeader    x-unique-id: <uuid>
-   BereqHeader    X-Forwarded-For: ...
-   BereqHeader    X-UA-Device: pc
-   BereqHeader    Host: test-host
-   BereqHeader    Surrogate-Capability: varnish=ESI/1.0
-   BereqHeader    X-Varnish: 1058444892
-   VCL_call       BACKEND_FETCH
-   VCL_return     fetch
-   BackendOpen    501 backend 12.34.56.78 8000 23.45.67.89 53480 reuse
-   FetchError     req.body read error: 104 (Connection reset by peer)
-   FetchError     backend write error: 104 (Connection reset by peer)
-   Timestamp      Bereq: 1616162471.629452 0.840149 0.840149
-   FetchError     Timed out reusing backend connection
-   BackendClose   501 backend close
-   Timestamp      Beresp: 1616162501.629545 30.840242 30.000093
-   Timestamp      Error: 1616162501.629548 30.840245 0.000002
-   BerespProtocol HTTP/1.1
-   BerespStatus   503
-   BerespReason   Backend fetch failed
-   BerespHeader   Date: Fri, 19 Mar 2021 14:01:41 GMT
-   BerespHeader   Server: Varnish
-   VCL_call       BACKEND_ERROR
-   BerespHeader   Content-Type: text/html; charset=utf-8
-   BerespHeader   Retry-After: 5
-   VCL_return     deliver
-   Storage        malloc Transient
-   Length         326
-   BereqAcct      1243 0 1243 0 0 0
-   End
```
Client part of the log:
```
*   << Request  >> 1058444891
-   Begin          req 1058444890 rxreq
-   Timestamp      Start: 1616162470.789225 0.000000 0.000000
-   Timestamp      Req: 1616162470.789225 0.000000 0.000000
-   VCL_use        reload_20210319_140100_14960
...
-   VCL_return     pass
-   VCL_call       HASH
-   VCL_return     lookup
-   VCL_call       PASS
-   VCL_return     fetch
-   Link           bereq 1058444892 pass
-   Storage        malloc Transient
-   FetchError     Connection reset by peer
-   FetchError     straight insufficient bytes
-   Timestamp      ReqBody: 1616162471.629420 0.840195 0.840195
-   Timestamp      Fetch: 1616162501.629593 30.840367 30.000172
-   RespProtocol   HTTP/1.1
-   RespStatus     503
-   RespReason     Backend fetch failed
-   RespHeader     Date: Fri, 19 Mar 2021 14:01:41 GMT
...
-   VCL_return     deliver
-   Timestamp      Process: 1616162501.629611 30.840386 0.000018
-   Filters        
-   RespHeader     Content-Length: 326
-   RespHeader     Connection: close
-   Timestamp      Resp: 1616162501.629634 30.840408 0.000022
-   ReqAcct        1156 0 1156 0 0 0
-   End            
```
There's a FetchError logged at the time the client closes the connection, but it waits for the first_byte_timeout (30s in this case).

## Your Environment
* Version used: varnish-6.5.1 revision 1dae23376bb5ea7a6b8e9e4b9ed95cdc9469fb64
* Operating System and version: debian buster with docker 20.10.2 and varnish with debian buster base image (still the latest version at time of writing this) with "net=host" and "ipc=host"
* Source of binary packages used: `deb https://packagecloud.io/varnishcache/varnish65/debian/ buster main`

**Comment by Dridi** [comment]
Thank you for your very thorough and helpful description of the problem!

**Comment by thetuxkeeper** [comment]
Thank you for the fast fix! Which release is planned to include your commit?

**Comment by Dridi** [comment]
I expect this to evntually land in 6.0.8 (cc @rezan) but can't really make have an ETA for the 6.5 series you are currently running, or even the recent 6.6 release.

If you aren't using any feature missing from our 6.0 LTS branch, I would recommend to switch to that one. As the name implies, that's our current long-term support branch. If you would be missing a feature from your current 6.5.1 installation, I'd be curious to know which one. I could at least confirm whether it could be elligible for a back-port, but that wouldn't be a commitment.

**Comment by thetuxkeeper** [comment]
I think we ran into #3535. We narrowed it down as a bug in varnish, but forgot to report it. Sorry for that!
Could have been discovered earlier (October 2020 was our migration to varnish 6).

**Comment by Dridi** [comment]
I think you can safely expect this and #3535 to be fixed in 6.0.8 in April.
